activerecord/lib/active_record/connection_adapters/connection_specification.rb

```diff
@@ -17,6 +17,37 @@ def to_hash
         @config.merge(name: @name)
       end
 
+      class ConnectionConfigurations < SimpleDelegator #:nodoc:
+        attr_accessor :root_level
+
+        def initialize(config)
+          super
+          @root_level = nil
+        end
+
+        def at(top_level)
+          normalized(top_level)
+        end
+
+        def normalized(key = @root_level)
+          config = key ? self[key] : self
+          config = config ? config.dup : {}
+
+          # if the configuration is a one level config, pushes that to be a two level config, with primary as key
+          if !config.key?("primary") && config.key?("adapter")
+            config = { "primary" => config }
+          end
+
+          if url = ENV["DATABASE_URL"]
+            config["primary"] ||= {}
```

Comment:
We've just spent lines establishing `"primary"`, surely it should have been made an empty hash by now?